server/src/main/java/org/elasticsearch/indices/TimeSeriesIdGeneratorService.java

```diff
@@ -0,0 +1,386 @@
+
+    /**
+     * Build the {@link TimeSeriesIdGenerator} async from the cluster state
+     * update thread. Creating this will queue a task to build the generator
+     * on the separate thread but return immediately. Callers to
+     * {@link #generator()} race that queued task. If they win they will
+     * build the {@link TimeSeriesIdGenerator} and if they lose they'll return
+     * a cached copy.
+     */
+    private static class AsyncValue extends Value {
+        private final LazyInitializable<TimeSeriesIdGenerator, RuntimeException> lazy;
+
+        private AsyncValue(long mappingVersion, LazyInitializable<TimeSeriesIdGenerator, RuntimeException> lazy) {
+            super(mappingVersion);
+            this.lazy = lazy;
+        }
+
+        AsyncValue(
+            long mappingVersion,
+            Function<IndexMetadata, TimeSeriesIdGenerator> buildTimeSeriesIdGenerator,
+            ExecutorService executor,
+            IndexMetadata indexMetadata
+        ) {
+            /*
+             * This closes over indexMetadata and keeps a reference to it
+             * for as long as the AsyncValue lives which is ok. It isn't the
+             * only thing with such a reference.
+             */
+            this(mappingVersion, new LazyInitializable<>(() -> buildTimeSeriesIdGenerator.apply(indexMetadata)));
+            executor.execute(new AbstractRunnable() {
+                @Override
+                protected void doRun() throws Exception {
+                    generator();
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    /*
+                     * We failed to build the time series id generator which sad,
+                     * but we don't have to make a ton of noise about it because
+                     * when someone goes to use it they'll attempt to build it
+                     * again. If *they* fail then it'll throw an exception to
+                     * the caller which'll get reported back over http.
+                     */
+                    logger.debug(
+                        new ParameterizedMessage("error building timeseries id generator for {} async", indexMetadata.getIndex()),
+                        e
+                    );
+                }
+            });
+        }
+
+        @Override
+        TimeSeriesIdGenerator generator() {
+            return lazy.getOrCompute();
+        }
+
+        @Override
+        Value withMappingVersion(long newMappingVersion) {
+            return new AsyncValue(newMappingVersion, lazy);
+        }
+    }
+
+    /**
+     * Key for deduplicating mappings. In an ideal world we'd just use the
+     * mapping's {@link CompressedXContent} but {@link CompressedXContent#equals(Object)}
+     * will try to decompress the mapping if the crc matches but the compressed bytes
+     * don't. That's wasteful for us - probably for everyone. If the crc and compressed
+     * bytes match that's a match.
```

Comment:
It's wasteful specifically because if we don't match we need to decompress the bytes anyway. Since that's slow we don't want to decompress a second time. I'm fairly sure it'd be safe to modify the `equals` implementation in `CompressedXContent` but I don't want to tempt fate in this PR.